daemons/pacemakerd/pacemakerd.c

```diff
@@ -553,6 +563,50 @@ pcmk_ipc_created(qb_ipcs_connection_t * c)
     crm_trace("Connection %p", c);
 }
 
+static void
+pcmk_handle_ping_request( crm_client_t *c, xmlNode *msg, uint32_t id)
+{
+        const char *value = NULL;
+        xmlNode *ping = NULL;
+        xmlNode *reply = NULL;
+        time_t pinged = time(NULL);
+        const char *from = crm_element_value(msg, F_CRM_SYS_FROM);
+
+        /* Pinged for status */
+        crm_trace("Pinged from %s.%s",
+                  crm_element_value(msg, F_CRM_ORIGIN),
+                  from?from:"unknown");
+        ping = create_xml_node(NULL, XML_CRM_TAG_PING);
+        value = crm_element_value(msg, F_CRM_SYS_TO);
+        crm_xml_add(ping, XML_PING_ATTR_SYSFROM, value);
+        crm_xml_add(ping, XML_PING_ATTR_PACEMAKERDSTATE, pacemakerd_state);
+        crm_xml_add_ll(ping, XML_ATTR_TSTAMP, (long long) pinged);
+        crm_xml_add(ping, XML_PING_ATTR_STATUS, "ok");
+        reply = create_reply(msg, ping);
+        free_xml(ping);
+        if (reply) {
+            if (crm_ipcs_send(c, id, reply, crm_ipc_server_event) <= 0) {
+                crm_err("Failed sending ping-reply");
+            }
+            free_xml(reply);
+        } else {
+            crm_err("Failed building ping-reply");
+        }
+        /* just proceed state on sbd pinging us */
+        if (from && strstr(from, "sbd")) {
```

Comment:
I'm thinking we should exit in the sbd + reply==NULL case. That should only be possible when out of memory.